src/Geometry/CubeGenerator.h

```diff
@@ -0,0 +1,149 @@
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ * @section DESCRIPTION
+ **/
+#ifndef CUBEGENERATOR_H
+#define CUBEGENERATOR_H
+
+#include "utils/logger.h"
+#include "utils/args.h"
+#include "MeshReader.h"
+
+#include <omp.h>
+
+#include <algorithm>
+#include <cstring>
+#include <map>
+#include <utility>
+#include <vector>
+
+#include <iostream>
+
+typedef int t_vertex[3];
+
+// changed Vertex to CubeVertex due to same struct naming here and in another part of SeisSol
+struct CubeVertex {
+  t_vertex v;
+
+  bool operator<(const CubeVertex& other) const {
+    return (v[0] < other.v[0]) || ((v[0] == other.v[0]) && (v[1] < other.v[1])) ||
+           ((v[0] == other.v[0]) && (v[1] == other.v[1]) && (v[2] < other.v[2]));
+  }
+};
+
+// Index of the vertices of a tetraedra in a cube
+// even/odd, index of the tetrahedra, index of vertex, offset of the vertices in x/y/z
+static const t_vertex TET_VERTICES[2][5][4] = {{{{0, 0, 0}, {1, 0, 0}, {0, 1, 0}, {0, 0, 1}},
+                                                {{1, 0, 0}, {0, 1, 0}, {1, 1, 1}, {1, 1, 0}},
+                                                {{1, 0, 0}, {1, 1, 1}, {0, 0, 1}, {1, 0, 1}},
+                                                {{0, 1, 0}, {0, 1, 1}, {0, 0, 1}, {1, 1, 1}},
+                                                {{1, 0, 0}, {0, 1, 0}, {0, 0, 1}, {1, 1, 1}}},
+                                               {{{0, 0, 0}, {0, 1, 0}, {0, 1, 1}, {1, 1, 0}},
+                                                {{0, 0, 0}, {1, 1, 0}, {1, 0, 1}, {1, 0, 0}},
+                                                {{0, 0, 0}, {1, 0, 1}, {0, 1, 1}, {0, 0, 1}},
+                                                {{1, 1, 0}, {1, 0, 1}, {1, 1, 1}, {0, 1, 1}},
+                                                {{0, 0, 0}, {1, 1, 0}, {0, 1, 1}, {1, 0, 1}}}};
+
+static const int TET_SIDE_NEIGHBORS[2][5 * 4] = {
+    {3, 3, 3, 0, 1, 3, 0, 2, 2, 2, 2, 1, 0, 1, 3, 1, 3, 0, 0, 2},
+    {2, 3, 0, 1, 1, 3, 3, 2, 2, 1, 1, 0, 0, 3, 2, 1, 2, 0, 0, 1}};
+
+static const int TET_SIDE_ORIENTATIONS[2][5 * 4] = {
+    {2, 0, 2, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 0},
+    {0, 1, 0, 0, 0, 1, 0, 2, 0, 0, 0, 2, 0, 0, 0, 0, 0, 0, 0, 0}};
+
+namespace seissol::geometry {
+
+struct CubeGeneratorParameters {
+  unsigned int cubeMinX;
+  unsigned int cubeMaxX;
+  unsigned int cubeMinY;
+  unsigned int cubeMaxY;
+  unsigned int cubeMinZ;
+  unsigned int cubeMaxZ;
+  unsigned cubeX;
+  unsigned cubeY;
+  unsigned cubeZ;
+  unsigned cubePx;
+  unsigned cubePy;
+  unsigned cubePz;
+  double cubeS;
+  double cubeSx;
+  double cubeSy;
+  double cubeSz;
+  double cubeTx;
+  double cubeTy;
+  double cubeTz;
+};
+
+class CubeGenerator : public seissol::geometry::MeshReader {
+  public:
+  CubeGenerator(int rank,
```

Comment:
As a suggestion, how about passing a SeisSol MPI object here directly?

Reply:
I used the existing `Geometry/NetcdfReader.cpp` and `Geometry/PUMLReader.cpp` files as a rough 'template' when I created the `CubeGenerator` class. There, we also don't pass a SeisSol MPI object.
If an MPI object is the preferred implementation choice, I can adjust the constructor such that we pass a SeisSol MPI object directly.